packages/multi-test/src/wasm.rs

```diff
@@ -410,31 +534,64 @@ pub struct WasmOps {
 }
 
 impl WasmOps {
-    pub fn commit<C>(self, router: &mut WasmRouter<C>)
-    where
-        C: Clone + fmt::Debug + PartialEq + JsonSchema,
-    {
-        self.new_contracts.into_iter().for_each(|(k, v)| {
-            router.contracts.insert(k, v);
+    pub fn commit(self, committable: &mut dyn WasmCommittable) {
+        committable.apply_ops(self)
+    }
+}
+
+pub trait WasmCommittable {
+    fn apply_ops(&mut self, ops: WasmOps);
+}
+
+impl<C> WasmCommittable for WasmRouter<C>
+where
+    C: Clone + fmt::Debug + PartialEq + JsonSchema,
+{
+    fn apply_ops(&mut self, ops: WasmOps) {
+        ops.new_contracts.into_iter().for_each(|(k, v)| {
+            self.contracts.insert(k, v);
         });
-        self.contract_diffs.into_iter().for_each(|(k, ops)| {
-            let storage = router.contracts.get_mut(&k).unwrap().storage.as_mut();
+        ops.contract_diffs.into_iter().for_each(|(k, ops)| {
+            let storage = self.contracts.get_mut(&k).unwrap().storage.as_mut();
             ops.commit(storage);
         });
     }
 }
 
+impl<'a, C> WasmCommittable for WasmCache<'a, C>
+where
+    C: Clone + fmt::Debug + PartialEq + JsonSchema,
+{
+    fn apply_ops(&mut self, ops: WasmOps) {
+        ops.new_contracts.into_iter().for_each(|(k, v)| {
+            self.state.contracts.insert(k, v);
+        });
+        ops.contract_diffs.into_iter().for_each(|(k, ops)| {
+            match self.state.get_contract_mut(self.parent_contracts, &k) {
+                Some(contract) => ops.commit(contract.storage),
+                None => panic!("No contract at {}, but applying diff", k),
+            }
+        });
+    }
+}
+
 impl<'a, C> WasmCache<'a, C>
 where
     C: Clone + fmt::Debug + PartialEq + JsonSchema,
 {
     fn new(router: &'a WasmRouter<C>) -> Self {
         WasmCache {
             router,
-            state: WasmCacheState {
-                contracts: HashMap::new(),
-                contract_diffs: HashMap::new(),
-            },
+            parent_contracts: router,
+            state: WasmCacheState::new(),
+        }
+    }
+
+    pub fn cache(&self) -> WasmCache<C> {
```

Comment:
I started reviewing this file, and this is a little off the top of my head at this point, but:

I _guess_ we need a multi-level (contract state) cache, to separate "server" contracts state from "client" contracts states. This, to mimic the transactional nature of the different contracts in the SDK.

Impressive piece of work, by the way.

Reply:
Before this PR we had a one-level cache in `App.execute_multi`. That is, all the messages are run together in a cache, and any error will cause the state changes to rollback. They will only be committed to the "real" storage if all return a success.

With submessages, each submessage will need it's own cache. They can write state and then fail, and then reply could catch the error and turn it to a success. That would mean the "parent contract" writes should be committed but the submessage writes not. These semanatics are described here: https://github.com/CosmWasm/cosmwasm/blob/main/SEMANTICS.md#submessages

This leads to the need of an arbitrarily multi-level caching technique. I feel my approach in this PR is a bit brute force and could probably use a bigger refactor to only use one Storage (rather than multiple HashMaps), and unify a lot on the caching in the end, but that would be much more work than needed to get a 0.7.0 release out. This package really should get some more refactors when we have capacity.